Update destructive.py

Shivansh-Game opened the pull request:
Fixes #3459 

I'd like to propose a fix for this issue. As @ryanavella pointed out, the NLTKWordTokenizer currently doesn't treat the em-dash (—) as a separate punctuation mark, which causes it to incorrectly merge adjacent words (e.g., food—which).

This is because the em-dash is missing from the PUNCTUATION regex list in nltk/tokenize/destructive.py.

The Fix
I've added a new rule to the PUNCTUATION list to specifically handle the em-dash by padding it with spaces.

Here is the diff for the change in nltk/tokenize/destructive.py:

```diff
--- a/nltk/tokenize/destructive.py
+++ b/nltk/tokenize/destructive.py
@@ -87,6 +87,7 @@
             re.compile(r"\.{2,}", re.U),
             r" \g<0> ",
         ),  # See https://github.com/nltk/nltk/pull/2322
         (re.compile(r"[;@#$%&]"), r" \g<0> "),
+        (re.compile(r"\u2014", re.U), r" \g<0> "), # Handle em-dashes
         (
             re.compile(r'([^\.])(\.)([\]\)}>"\']*)\s*$'),
             r"\1 \2\3 ",
```

Test Results
With this fix, the tokenizer now correctly handles the example provided in the issue:

Python

import nltk
s = "The food—which was delicious—reminded me of home."
Before:

Python

>>> nltk.word_tokenize(s)
['The', 'food—which', 'was', 'delicious—reminded', 'me', 'of', 'home', '.']
After:

Python

>>> nltk.word_tokenize(s)
['The', 'food', '—', 'which', 'was', 'delicious', '—', 'reminded', 'me', 'of', 'home', '.']
Regarding the pos_tag Issue
This tokenization fix also resolves the root cause of the incorrect pos_tag output. The tagger was previously receiving "garbage" tokens (food—which) and making incorrect guesses.

With the corrected tokens, the pos_tag output is now logical:

Python

>>> nltk.pos_tag(nltk.word_tokenize(s))
[('The', 'DT'), ('food', 'NN'), ('—', 'NNP'), ('which', 'WDT'), ('was', 'VBD'), 
 ('delicious', 'JJ'), ('—', 'NNP'), ('reminded', 'VBD'), ('me', 'PRP'), 
 ('of', 'IN'), ('home', 'NN'), ('.', '.')]


The tagger correctly identifies all the words (food, which, delicious, etc.). It appears to default to NNP (Proper Noun) for the lone — token, which is a reasonable guess from the statistical model given it's likely an unknown token in that context. The core tokenization problem is solved.

I have tested this locally and am happy to open a Pull Request.

Shivansh-Game committed:
Update destructive.py

cclauss commented on nltk/tokenize/destructive.py:
```diff
@@ -85,7 +85,8 @@ class NLTKWordTokenizer(TokenizerI):
         ),  # See https://github.com/nltk/nltk/pull/2322
-        (re.compile(r"[;@#$%&]"), r" \g<0> "),
+        (re.compile(r"[;@#$%&]"), r" \g<0> 
+        (re.compile(r"\u2014", re.U), r" \g<0> "), # \u2014 handles emdashes
```
Readability: [`ruff rule FURB167`](https://docs.astral.sh/ruff/rules/regex-flag-alias)
```suggestion
        (re.compile(r"\u2014", re.UNICODE), r" \g<0> "),  # \u2014 handles emdashes
```

Shivansh-Game committed:
Update nltk/tokenize/destructive.py

Co-authored-by: Christian Clauss <cclauss@me.com>

Shivansh-Game committed:
Update destructive.py

Shivansh-Game commented:
Hello maintainers, the fix is causing a test to fail in lm.test_models.py, only on the macos-3.11/3.10 build. This is my first contribution, so I'm not confident enough in my ability to fix an issue like this. Could a maintainer please take a look?

purificant commented:
Triggered test re-run

ekaf reviewed (changes requested):
Goal: Pad em dashes so tokens like “food—which” are split correctly.

Problem: The pattern proposed here uses raw string r"\u2014"; this matches the literal backslash-u sequence, not the em dash. re.UNICODE doesn’t fix that.
- Fix: Use "\u2014" (non-raw), the literal "—", or a class "[\u2014]".
- Optional enhancement: Broaden to "[\u2012-\u2015]" to cover ‒ – — ―.

Placement in PUNCTUATION is appropriate; existing whitespace normalization will handle added spaces.

Please add unit tests for em dash (and optionally other dash characters) to prevent regressions.

Shivansh-Game committed:
Update destructive.py

Shivansh-Game committed:
Merge branch 'nltk:develop' into fix-em-dash-3459

Shivansh-Game committed:
Update test_tokenize.py

Shivansh-Game commented:
Thanks for pointing that out @ekaf, I've fixed it and broadened the scope to cover the other dash characters you suggested. I've also added the unit tests for all of the new dash characters.

ekaf commented:
@Shivansh-Game, please note that CI fails already on pre-commit. This is because you need to run both _git add_ and _git commit_ again whenever black or isort modify any of your files.

Shivansh-Game committed:
Chore: Apply black and isort formatting

Shivansh-Game commented:
Hi @ekaf, I've just pushed the new commit with the pre-commit formatting applied.
This was my first time doing the full git workflow, thank you walking me through that. o7

ekaf commented:
@Shivansh-Game, to debug the latest CI error, you could try to run "pytest  -v nltk/test/unit/test_tokenize.py" on your local machine before your next commit.

Shivansh-Game committed:
Fix: unit tests for dash tokenization

Shivansh-Game commented:
Hey @ekaf, thanks for the advice and patience, I ran all the tests locally after fixing the errors in the previous commit. It should work now.

ekaf reviewed (changes requested):
Thanks @Shivansh-Game, this looks good., but: 
 - Horizontal bar (U+2015) is not actually tested. The test labeled “Horizontal Bar (\u2015)” uses the em dash glyph (—, U+2014) again in both the input string and expected tokens. The horizontal bar character is ― (slightly longer visually). 
- (Optionally) Add one regression guard for hyphen-minus: e.g. assert word_tokenize("state-of-the-art") == ["state-of-the-art"] (or the current canonical expected output) to be sure behavior is unchanged.

Shivansh-Game committed:
Fixed unit test for horizontal bar + added a regression guard for hyphen-minus

Shivansh-Game commented:
Hey @ekaf, Thanks for pointing that out. I've fixed that unit test and added the regression guard. I've tested it locally and it should test for horizontal bars properly now and pass the CI.

ekaf reviewed (approved):
LGTM, thanks @Shivansh-Game!

Shivansh-Game commented:
Hey, I just wanted to ask if there's anything I should do on my end that I might have overlooked since this pull request has been approved for a while now.
